doc/design/containerd/port-layer-integration.md

```diff
@@ -0,0 +1,276 @@
+
+	Resume(context.Context, *Container) error
+
+	List(context.Context) ([]*Container, error)
+
+	Load(ctx context.Context, id string) (*Container, error)
+
+	Delete(context.Context, *Container) error
+
+	Start(context.Context, *Container) error
+
+	StartProcess(context.Context, *Container, StartProcessOpts) (Process, error)
+
+	SignalProcess(ctx context.Context, c *Container, id string, sig os.Signal) error
+
+	DeleteProcess(ctx context.Context, c *Container, id string) error
+
+}
+
+```
+*Image Pull*
+
+Image pull will be handled by the "Distribution" subsystem, which it is to be assumed operates similarly to VIC imagec component. Given a URI, it should be able to pull down any images that don’t already exist in a local cache. 
+
+The question of whether snapshots are created as output of image pull or whether these are created as output of container create is one that gets to the heart of whether the output of image pull needs to be cached at all. At present this isn’t clear, but it would be preferable for it to be the latter.
+
+Image metadata will need to persist beyond the lifespan of the VIC appliance and as such will need to be written into vSphere.
+
+*Container Create*
+
+Assuming the image snapshots are already created, an additional RW layer needs to be created, assuming the config doesn’t have readonly=true. 
+
+There is then a configuration pre-processing that needs to happen where the config.json is generated for the container. This will encapsulate a mixture of the image spec, containerd command-line arguments, constants set for the installation and possibly environmental factors.
+
+There will then need to be a validation step of the generated config to make sure that it makes sense. For example, referenced networks and storage need to exist. If the config passes validation, create() is called on the executor interface, which would propagate to the Port Layer. A VMConfigSpec will then be generated and a VM will be created. This is the equivalent of runc create (where a dummy process is created and gets a PID).
+
+*Container Run*
+
+Container run should operate in a very similar manner to the way it does today in that the VM is powered on and a handshake established with the Tether.
+
+The question is how we handle TTY/attach and streaming of stdout and stderr. In the current version of containerd, the containerd-shim maintains a Fifo on the filesystem that can be reconnected to if containerd shuts down and restarts.
+
+<TBD>
+
+**Questions**
+
+There are a number of questions that still need to be worked out. Here’s are the most obvious:
+
+*Do we need to create a process per container in the VIC appliance to fit with the executor model?*
+
+Both runc and the containerd shim are a process-per-container model. Our Port Layer is a long-running service. We will need some kind of shim layer on the Port Layer to mate with the executor interface anyway. Does having entries in /proc for each container help in any way?
```

Comment:
there is a way to implement runtime plugin that can recover the state and fake up all running shims.